packages/web/app/src/lib/preflight-sandbox/graphiql-plugin.tsx

```diff
@@ -165,23 +170,47 @@ export function usePreflightScript(args: {
     'hive:laboratory:isPreflightScriptEnabled',
     false,
   );
-  const [environmentVariables, setEnvironmentVariables] = useLocalStorage(
-    'hive:laboratory:environment',
-    '',
-  );
+
+  // ------------
+  // Result State
+  // ------------
+  //
+  // todo: Probably better to store the result as a single JSON object value.
+  // Use a proper versioned schema with codecs for coercing, defaults, validation, etc.
+  //
+  // todo: Improve `useLocalStorage` by allowing passing a codec? Then we can co-locate
+  // the codec with the data and have it applied transparently, use a decoded value
+  // for the default, etc. ?
+
+  const [environmentVariables, setEnvironmentVariables] = useLocalStorage('hive:laboratory:environment', '{}'); // prettier-ignore
   const latestEnvironmentVariablesRef = useRef(environmentVariables);
-  useEffect(() => {
-    latestEnvironmentVariablesRef.current = environmentVariables;
-  });
+  useEffect(() => { latestEnvironmentVariablesRef.current = environmentVariables; }); // prettier-ignore
+  const decodeEnvironmentVariables = (encoded: string) => safeParseJSON<Result['environmentVariables']>(encoded) ?? {}; // prettier-ignore
+
+  const [headers, setHeaders] = useLocalStorage('hive:laboratory:headers', '[]');
+  const latestHeadersRef = useRef(headers);
+  useEffect(() => { latestHeadersRef.current = headers; }); // prettier-ignore
+  const decodeHeaders = (encoded: string) => safeParseJSON<Result['headers']>(encoded) ?? [];
+
+  const decodeResult = (): Result => {
+    return {
+      environmentVariables: decodeEnvironmentVariables(latestEnvironmentVariablesRef.current), // prettier-ignore
+      headers: decodeHeaders(latestHeadersRef.current),
+    };
+  };
+  // -----------
 
   const [state, setState] = useState<PreflightWorkerState>(PreflightWorkerState.ready);
   const [logs, setLogs] = useState<LogRecord[]>([]);
 
   const currentRun = useRef<null | Function>(null);
 
-  async function execute(script = target?.preflightScript?.sourceCode ?? '', isPreview = false) {
+  async function execute(
+    script = target?.preflightScript?.sourceCode ?? '',
+    isPreview = false,
+  ): Promise<Result> {
```

Comment:
My implementation approach boils down to having `execute` return an extensible set of data. In this PR it includes environment variables as before and, newly, headers.

The current code returns environment variables directly.